src/clr-angular/forms/password/password-container.ts

```diff
@@ -0,0 +1,108 @@
+import { LayoutService } from '../common/providers/layout.service';
+import { DynamicWrapper } from '../../utils/host-wrapping/dynamic-wrapper';
+import { ControlIdService } from '../common/providers/control-id.service';
+import { ClrLabel } from '../common/label';
+import { ClarityIcons } from '@clr/icons';
+import { ClrPassword } from './password';
+import { ClrShapeEye, ClrShapeEyeHide } from '@clr/icons/shapes/essential-shapes';
+import { FocusService } from '../common/providers/focus.service';
+import { ControlClassService } from '../common/providers/control-class.service';
+
+@Component({
+  selector: 'clr-password-container',
+  template: `
+    <ng-content select="label"></ng-content>
+    <label *ngIf="!label && addGrid()"></label>
+    <div class="clr-control-container" [ngClass]="controlClass()">
+      <div class="clr-input-wrapper">
+        <div class="clr-input-group" [class.clr-focus]="focus">
+          <ng-content select="[clrPassword]"></ng-content>
+          <clr-icon shape="eye" *ngIf="!show && clrToggle" class="clr-input-group-icon-action" (click)="toggle()"></clr-icon>
+          <clr-icon shape="eye-hide" *ngIf="show && clrToggle" class="clr-input-group-icon-action" (click)="toggle()"></clr-icon>
+        </div>
+        <clr-icon *ngIf="invalid" class="clr-validate-icon" shape="exclamation-circle"></clr-icon>
+      </div>
+      <ng-content select="clr-control-helper" *ngIf="!invalid"></ng-content>
+      <ng-content select="clr-control-error" *ngIf="invalid"></ng-content>
+    </div>
+    `,
+  host: {
+    '[class.clr-form-control]': 'true',
+    '[class.clr-row]': 'addGrid()',
+  },
+  providers: [IfErrorService, NgControlService, ControlIdService, ControlClassService, FocusService],
+})
+export class ClrPasswordContainer implements DynamicWrapper, OnDestroy {
+  subscriptions: Subscription[] = [];
+  invalid = false;
+  control: NgControl;
+  _dynamic = false;
+  show = false;
+  focus = false;
+  @Input() clrToggle = true;
+  @ContentChild(ClrLabel) label: ClrLabel;
+  @ContentChild(forwardRef(() => ClrPassword))
+  password: ClrPassword;
+
+  constructor(
+    private ifErrorService: IfErrorService,
+    @Optional() private layoutService: LayoutService,
+    @Optional() private controlClassService: ControlClassService,
+    public focusService: FocusService,
+    private renderer: Renderer2
+  ) {
+    this.subscriptions.push(
+      this.ifErrorService.statusChanges.subscribe(control => {
+        this.invalid = control.invalid;
+      })
+    );
+    this.subscriptions.push(
+      this.focusService.focusChange.subscribe(state => {
+        this.focus = state;
+      })
+    );
+    // Explicitly add eye for this component
+    ClarityIcons.add({
+      eye: ClrShapeEye,
+      'eye-hide': ClrShapeEyeHide,
+    });
+  }
+
+  toggle() {
+    this.show = !this.show;
+    this.renderer.setProperty(this.password.el.nativeElement, 'type', this.show ? 'text' : 'password');
```

Comment:
I know it's a pain to move something that small, but this should happen in the `ClrPassword` directive, not in the container. And that way you wouldn't need to `@ContentChild` it, you wouldn't have cyclic dependencies that you're solving with `forwardRef`, etc.

Feel free to just declare a provider with `useFactory: () => new BehaviorSubject<boolean>()` to communicate the show/hide if you want to remain concise.

Reply:
done.